### ISSUE TITLE
Finalize the keychain and account management model

### ISSUE THREAD

**Issue by shea256** [blockstack/blockstack-browser#131]
Let's have a discussion about how to best handle our keychain and account management needs. I wrote some notes about the available keychain types, as well as a proposed data model.

### Keychain Structure

Right now we have two options for keychains:

- Enumerated keys (HD keys)
  - https://github.com/blockstack/keychain-manager-js
- Hash-based keys
  - https://github.com/blockstack/blockstack-keychains-js

We have libraries for both but they each have their advantages and we may need both for different things.

### Data Model

- Master Keychain / Identity
  - Actions
    - Create keychain
    - Get keychain
    - Delete keychain
  - Subobjects
    - Identity keychain
    - Name keychain
    - Token keychain (for BTC, ETH, etc)
- Child keychain
  - Subobjects
    - Master key (extended private key)
    - Master key ID (extended public key)
    - Keypairs
- Keypairs
  - Subobjects
    - Key (private key)
    - Key ID (public key)
    - Address

**Comment by shea256** [comment]
### Keychain / Account Assumptions

- A browser account *is* a keychain
- A user can have one keychain at a time
- A user's keychain can be created, restored and deleted
- The keychain comes a recovery phrase
- The keychain can be encrypted with a password
- The keychain comes with three hardened child keychains

**Comment by shea256** [comment]
**Token keychain:** for sending bitcoin transactions
**Name keychain:** for sending blockstack name transactions
**Identity keychain:** for signing messages that can be used to authenticate into applications and for signing data payloads

**Comment by shea256** [comment]
Since we're going with HD Keys as the top-level master keychain, we'll need to define the paths for the identity keychain and the name keychain.

### Design

One option:

| Child Keychain | HD Path |
| --------------- | -------- |
| Bitcoin keychain (account 0) | m / 44' / 0' / 0' |
| Identity keychain (account 0) | m / 201' / 0' / 0' |
| Name keychain (account 0) | m / 202' / 0' / 0' |

The path for the bitcoin keychain means "this keychain is defined by BIP44 (`/ 44'`) and it's using Bitcoin (`/ 0'`) and it's the first account (`/ 0'`).

We could create a proposal for name keychains and say that the proposal number is 202 and Blockstack's number is 0.

Now, once again, I don't like this system because it's very dependent on circulated lists and it's not a great way to do collision avoidance.

Another thing we can do is the following:

| Child Keychain | HD Path |
| --------------- | -------- |
| Bitcoin keychain (account 0) | m / 44' / 0' / 0' |
| Identity keychain (account 0) | m / hash("blockstack-identity-0") |
| Name keychain (account 0) | m / hash("blockstack-naming-0") |

Basically, we'd only use HD Key derivation for the Bitcoin keychain and for the root keychain of the other two.

### Implementation

Here's how we would implement it if we're working with bitcoinjs-lib:

```js
import {randomBytes} from 'crypto'
import {HDNode} from 'bitcoinjs-lib'
const masterKeychain = HDNode.fromSeedBuffer(randomBytes(32))
const masterKeychainID = masterKeychain.neutered()
const bitcoinKeychain = masterKeychain.deriveHardened(44).deriveHardened(0)
const identityKeychain = masterKeychain.deriveHardened(201).deriveHardened(0)
const nameKeychain = masterKeychian.deriveHardened(202).deriveHardened(0)
```

**Comment by jackzampolin** [comment]
Looks like this got solved over here: https://github.com/blockstack/blockstack-browser/issues/330
